dashboard/app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -453,12 +461,24 @@ def sign_in_user(user)
     sign_in_and_redirect user
   end
 
+  def email_already_taken(user)
+    lookup_user = User.find_by_email_or_hashed_email(user.email)
+    return !!lookup_user
+  end
+
+  def auth_already_exists(auth_hash)
+    lookup_user = User.find_by_credential(type: auth_hash.provider, id: auth_hash.uid)
+    return !!lookup_user
+  end
+
   def allows_silent_takeover(oauth_user, auth_hash)
     return false unless auth_hash.provider.present?
     return false unless AuthenticationOption::SILENT_TAKEOVER_CREDENTIAL_TYPES.include?(auth_hash.provider.to_s)
     return false if oauth_user.persisted?
 
-    lookup_user = User.find_by_email_or_hashed_email(oauth_user.email)
+    verified_email_credentials = AuthenticationOption::TRUSTED_EMAIL_CREDENTIAL_TYPES - [AuthenticationOption::EMAIL]
+
+    lookup_user = AuthenticationOption.where(credential_type: verified_email_credentials).find_by(hashed_email: User.hash_email(oauth_user.email))&.user || User.where(hashed_email: User.hash_email(oauth_user.email)).where(provider: verified_email_credentials).first
```

Comment:
[i know this is a draft, so excuse the nit] we should break this into multiple lines for readability and so that we only need to look up the `User` once